store/src/errors.rs

```diff
@@ -1,64 +1,198 @@
+        Self::StateError(Box::new(value))
+    }
+}
+
+// Genesis errors
+// =================================================================================================
+
+#[derive(Debug, Error)]
+pub enum GenesisError {
+    #[error("Apply block failed: {0}")]
+    ApplyBlockFailed(String),
+    #[error("Failed to read genesis file \"{genesis_filepath}\": {error}")]
+    FailedToReadGenesisFile {
+        genesis_filepath: String,
+        error: io::Error,
+    },
+    #[error("Failed to deserialize genesis file: {0}")]
+    GenesisFileDeserializationError(DeserializationError),
+    #[error("Block header in store doesn't match block header in genesis file. Expected {expected_genesis_header:?}, but store contained {block_header_in_store:?}")]
+    GenesisBlockHeaderMismatch {
+        expected_genesis_header: Box<BlockHeader>,
+        block_header_in_store: Box<BlockHeader>,
+    },
+    #[error("Malconstructed genesis state: {0}")]
+    MalconstructedGenesisState(MerkleError),
+    #[error("Retrieving genesis block header failed: {0}")]
+    SelectBlockHeaderByBlockNumError(Box<DbError>),
+}
+
+// State errors
+// =================================================================================================
+
+#[derive(Error, Debug)]
 pub enum StateError {
-    ConcurrentWrite,
+    #[error("Database error: {0}")]
+    DatabaseError(#[from] DbError),
+    #[error("Conversion error: {0}")]
+    ConversionError(#[from] ConversionError),
+    #[error("Database doesnt have any block header data")]
     DbBlockHeaderEmpty,
-    DigestError(ParseError),
-    DuplicatedNullifiers(Vec<RpoDigest>),
-    InvalidAccountId,
-    MissingAccountHash,
-    MissingAccountId,
-    MissingNoteHash,
-    NewBlockInvalidAccountRoot,
+    #[error("Failed to get MMR peaks for forest ({forest}): {error}")]
+    FailedToGetMmrPeaksForForest { forest: usize, error: MmrError },
+    #[error("Chain MMR forest expected to be 1 less than latest header's block num. Chain MMR forest: {forest}, block num: {block_num}")]
+    IncorrectChainMmrForestNumber { forest: usize, block_num: u32 },
+}
+
+#[derive(Error, Debug)]
+pub enum StateInitializationError {
+    #[error("Database error: {0}")]
+    DatabaseError(#[from] DbError),
+    #[error("Conversion error: {0}")]
+    ConversionError(#[from] ConversionError),
+    #[error("Failed to create nullifiers tree: {0}")]
+    FailedToCreateNullifiersTree(MerkleError),
+    #[error("Failed to create accounts tree: {0}")]
+    FailedToCreateAccountsTree(MerkleError),
+}
+
+#[derive(Error, Debug)]
+pub enum StateSyncError {
+    #[error("Database error: {0}")]
+    DatabaseError(#[from] DbError),
+    #[error("Block database is empty")]
+    BlockDbIsEmpty,
+    #[error("Failed to get MMR delta: {0}")]
+    FailedToGetMmrDelta(MmrError),
+}
+
+// Block applying errors
+// =================================================================================================
+
+#[derive(Error, Debug)]
+pub enum ApplyBlockError {
```

Comment:
Let's create an issue to revisit variants of this error. I think we can consolidate and normalize them a bit better than how they are now.

Reply:
@bobbinth thank you. I've created an issue: https://github.com/0xPolygonMiden/miden-node/issues/214
Please, help me to fill it by providing more context.